File: handlerclaw/lib/features/finances/data/dto/account_type_dto.dart
import 'package:handlerclaw/features/finances/data/dto/account_dto.dart';

class AccountTypeDto {
  final String id;
  final String name;
  final String category;
  final double currentTotalAmount;
  final int accountCount;
  final List<AccountDto> accounts;

  AccountTypeDto({
    required this.id,
    required this.name,
    required this.category,
    required this.currentTotalAmount,
    required this.accountCount,
    required this.accounts,
  });

  factory AccountTypeDto.fromJson(Map<String, dynamic> json) {
    return AccountTypeDto(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      category: json['category'] ?? '',
      currentTotalAmount: (json['current_total_amount'] ?? 0).toDouble(),
      accountCount: json['account_count'] ?? 0,
      accounts: (json['accounts'] as List?)
              ?.map((e) => AccountDto.fromJson(e))
              .toList() ??
          [],
    );
  }
}

class AccountCategoryDto {
  final String name;
  final int totalAccountTypes;
  final double totalAmount;

  AccountCategoryDto({
    required this.name,
    required this.totalAccountTypes,
    required this.totalAmount,
  });

  factory AccountCategoryDto.fromJson(Map<String, dynamic> json) {
    return AccountCategoryDto(
      name: json['name'] ?? '',
      totalAccountTypes: json['total_account_types'] ?? 0,
      totalAmount: (json['total_amount'] ?? 0).toDouble(),
    );
  }
}

class AccountTypesPaginatedDto {
  final List<AccountTypeDto> accountTypes;
  final List<AccountCategoryDto> categories;

  AccountTypesPaginatedDto({
    required this.accountTypes,
    required this.categories,
  });

  factory AccountTypesPaginatedDto.fromJson(Map<String, dynamic> json) {
    return AccountTypesPaginatedDto(
      accountTypes: (json['account_types'] as List?)
              ?.map((e) => AccountTypeDto.fromJson(e))
              .toList() ??
          [],
      categories: (json['categories'] as List?)
              ?.map((e) => AccountCategoryDto.fromJson(e))
              .toList() ??
          [],
    );
  }
}

File: api/src/modules/(jwt-auth)/finances/account-types/application/dtos/account-type.dto.ts
import type { AccountType } from "../../../../../../lib/generated/prisma/client.js";
import type { PaginationMetaDto } from "../../../../../../lib/types/pagination-meta-dto.js";
import type { IAccountType } from "../../domain/entities/account-type.entity.js";
import type { AccountCategory } from "../../../../../../lib/generated/prisma/enums.js";
import type { FinanceOverviewAccountItem } from "../../../overview/application/dtos/finance-overview.dto.js";

/**
 * Shared Data Structures
 */
export type AccountTypeCategorySummary = {
  name: string;
  total_account_types: number;
  total_amount: number;
};

export type AccountTypeWithMetrics = AccountType & {
  current_total_amount: number;
  account_count: number;
  accounts: {
    id: string;
    name: string;
    category: AccountCategory;
    current_amount: number;
  }[];
};

export type FindAllAccountTypesResult = {
  accountTypes: AccountTypeWithMetrics[];
  total: number;
  categories: AccountTypeCategorySummary[];
};

/**
 * Input Data Contracts
 */
export type CreateAccountTypeData = {
  name: string;
  category: AccountCategory;
};

export type UpdateAccountTypeData = Partial<CreateAccountTypeData>;

export type CreateAccountTypeRequest = {
  name: string;
  category: AccountCategory;
};

export type UpdateAccountTypeRequest = {
  name?: string | undefined;
  category?: AccountCategory | undefined;
};

export type GetAccountTypesQuery = {
  page: number;
  per_page: number;
  search?: string | undefined;
};

export type IAccountTypeWithMetrics = IAccountType & {
  current_total_amount: number;
  account_count: number;
  accounts: FinanceOverviewAccountItem[];
};

/**
 * Response Contracts
 */
export type CreateAccountTypeResponse = {
  account_type: IAccountType;
};

export type GetAccountTypesResponse = {
  categories: AccountTypeCategorySummary[];
  account_types: IAccountTypeWithMetrics[];
  meta: PaginationMetaDto;
};

export type GetAccountTypeDetailResponse = {
  account_type: IAccountType;
};

export type UpdateAccountTypeResponse = {
  account_type: IAccountType;
};
